larq_compute_engine/mlir/transforms/prepare_patterns.td

```diff
@@ -110,3 +113,70 @@ def : Pat<(TF_Conv2DOp:$output
            (ConstFloatValueIsOne $pad_values),
            (SamePadding $paddings, $input, $output, $strides)],
           (addBenefit 90)>;
+
+// Match a binary convolution with unipolar (+1/0) inputs, padded (if
+// applicable) with zeroes. The weights are bipolar (+1/-1). We use the
+// following trick to perform this computation using the bconv2d op (which only
+// supports bipolar inputs).
+//     Suppose `a` is a unipolar (+1/0) activation and `w` is a bipolar (+1/-1)
+// weight. If we define `b` to be 1 if `a` is 1 and `-1` if `a` is 0, then
+// `a = 0.5 * b + 0.5`. This means that `a * w` can be computed with bipolar
+// multiplication followed by a non-binary multiplication and bias:
+//            `a * w = (0.5 * b + 0.5) * w = 0.5 * b * w + 0.5 * w`
+// Note that the bias `0.5 * w` is a constant at conversion-time. The
+// generalisation to more than a single activation/weight element is then:
+//         `dotprod(a, w) = 0.5 * dotprod(b, w) + 0.5 * reduce_sum(w)`
+// which is the form we use below.
+//     There is one final thing to note here. Because the unipolar input 1 is
+// represented by bipolar input 1 and unipolar input 0 by bipolar input -1, TF
+// zero-padding is equivalent to padding with -1 in LCE; however, the bconv2d op
+// only supports padding with 1. The solution is to multiply the (bipolar) input
+// by -1 and the weights by -1 -- this means that TF zero-padding is equivalent
+// to LCE bconv2d one-padding.
```

Comment:
Instead of the "multiplication by -1" argument, can you not do this directly?
Let `b=1` if `a=0` and `b=-1` if `a=1` (i.e. flipped from what you write here). Then `a = -0.5 * b + 0.5` (just flips the signs here). The rest of your computation is similar then:
`dotprod(a,w) = -0.5 * dotprod(b, w) + 0.5 * reduce_sum(w)`.
Its entirely the same as what you have but it avoid the extra "multiplication by -1" argument for zero-padding so it might be simpler to understand for the reader of the comment.

Reply:
I originally had something very similar to what you are suggesting here. It forces you to pattern match against a weird formulation of the Larq quantizer, e.g. `-0.5 * (threshold - input) + 0.5`, which is equivalent to having pushed through a -1, the only difference is that the -1 is 'baked in' to the quantiser definition.

Reply:
Incidentally, the multiplication by -1 isn't needed if you do unipolar + one-padding rather than unipolar + zero-padding.